server/src/main/scala/org/http4s/server/middleware/authentication/NonceF.scala

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.server.middleware.authentication
+
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+import org.http4s.crypto.unsafe.SecureRandom
+
+import java.math.BigInteger
+import java.util.Date
+
+private[authentication] class NonceF[F[_]](val created: Date, val nc: Ref[F, Int], val data: String)
+
+private[authentication] object NonceF {
+  val random = new SecureRandom()
+
+  private def getRandomData(bits: Int): String = new BigInteger(bits, random).toString(16)
```

Comment:
This is potentially a blocking call.  It's going to be fine for most cases, but have the same problems on exotic platforms as as https://github.com/typelevel/cats-effect/issues/2882.

Probably not urgent: it was that way before, and nobody noticed enough to report it.

Reply:
Yeah, this caught my eye too. I guess if it's okay we can just leave it, although @blast-hardcheese make sure to wrap this in an `F.delay` b/c generating random numbers is a side-effect.

Reply:
Arman brought up how it's difficult to do "right" in CE2, and that's true.  `F.delay` is appropriate for the vast majority of users, but `F.blocking` is required in more exotic cases.  In CE3, the interfaces are the same, so we can detect the exotic cases and block only where needed.  In CE2, we can also detect, but we'd need to lug around a `Blocker` just in case.

I'll see if I can draft a quick helper.  The logic will be mostly what we need for that CE3 ticket anyway.